main_test.go

```diff
@@ -16,16 +16,78 @@ package main
 
 import (
 	"context"
+	"fmt"
 	"math/rand"
 	"net"
 	"net/http"
 	"net/http/httptest"
+	"strings"
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/uuid"
 )
 
+func TestValidate(t *testing.T) {
+	tests := []struct {
+		desc      string
+		input     configInput
+		wantError string
+	}{
+		{
+			desc: "fails when config-scope has too many characters",
+			input: configInput{
+				xdsServerUri:      "example.com:443",
+				gcpProjectNumber:  123456789012345,
+				vpcNetworkName:    "thedefault",
+				ip:                "10.9.8.7",
+				zone:              "uscentral-5",
+				metadataLabels:    map[string]string{"k1": "v1", "k2": "v2"},
+				includeV3Features: true,
+				configScope:       strings.Repeat("a", 65),
+			},
+			wantError: "config-scope may only contain letters, numbers, and '-'. It must begin with a letter and must not exceed 64 characters in length",
+		},
+		{
+			desc: "fails when config-scope does not start with an alphabetic letter",
+			input: configInput{
+				xdsServerUri:      "example.com:443",
+				gcpProjectNumber:  123456789012345,
+				vpcNetworkName:    "thedefault",
+				ip:                "10.9.8.7",
+				zone:              "uscentral-5",
+				metadataLabels:    map[string]string{"k1": "v1", "k2": "v2"},
+				includeV3Features: true,
+				configScope:       "4foo",
+			},
+			wantError: "config-scope may only contain letters, numbers, and '-'. It must begin with a letter and must not exceed 64 characters in length",
+		},
+		{
+			desc: "fails when config-scope contains characters besides letters, numbers, and hyphens.",
+			input: configInput{
+				xdsServerUri:      "example.com:443",
+				gcpProjectNumber:  123456789012345,
+				vpcNetworkName:    "thedefault",
+				ip:                "10.9.8.7",
+				zone:              "uscentral-5",
+				metadataLabels:    map[string]string{"k1": "v1", "k2": "v2"},
+				includeV3Features: true,
+				configScope:       "h*x8",
+			},
+			wantError: "config-scope may only contain letters, numbers, and '-'. It must begin with a letter and must not exceed 64 characters in length",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			err := validate(test.input)
+			if test.wantError != fmt.Sprint(err) {
```

Comment:
I would use `err.Error()` instead of `fmt.Sprintf(err)`.

And the more common way to print an `err` here would be:
t.Fatalf("blah foo: %v", err)
https://tour.golang.org/methods/19

Reply:
Done.